### ISSUE TITLE
Suggestion: PS4 Texture swizzling/unswizzling

### ISSUE THREAD

**Issue by eikun79** [nesrak1/UABEA#297]
Recently, PS4 is also seeing an increasing number of games with Swizzle image files, 
such as Nintendo Switch
Can you help me convert PS4 Swizzle files to Unsweezle as below?

[NotoSansJP-Regular SDF Atlas-sharedassets0.assets-590.zip](https://github.com/nesrak1/UABEA/files/13667489/NotoSansJP-Regular.SDF.Atlas-sharedassets0.assets-590.zip)

Example) https://github.com/nesrak1/UABEA/issues/185

**Comment by nesrak1** [comment]
I'm not sure how close ps5 swizzling is to ps4, but some work was done for ps5 textures in the past, but I don't know where that fork is right now. I don't think it should be too terrible to implement that into the current version since switch swizzle support exists as a starting point, but I would need people to help test since I don't own a ps4.

**Comment by eikun79** [comment]
The game in the file I uploaded yesterday is il2cpp version 29, and there is another problem because the managed cannot be extracted.
(So far, working on PS4unity localization has produced some works... for SDF font application, monobehaviour (.txt) extraction is essential, but version 29 cannot be extracted...)

enreal engine games have a lot of experience working on swizzle images (dds to gnf) but I've only seen swizzle images of unity games twice so far,
As we move on to the latest 22~23 years, the image of Unity Swizzle seems to be increasing.

**Comment by fdwr** [comment]
What a peculiar case. If you apply the same Morton-order nested tiling (up to 8x8 pixel blocks in order: x:2 y:2 x:4 y:4 x:8 y:8) as you would for other PS4 textures ...

![Image](https://github.com/user-attachments/assets/e6fec49c-e627-4048-8ee2-628e8484f784)

...then you get this:

![Image](https://github.com/user-attachments/assets/ce5ecbf3-50e6-445d-bb38-2daefb8e1731)

But if you just horizontally flip the texture before unswizzling (and horizontally flip again after), then you get this:

![Image](https://github.com/user-attachments/assets/7c28173a-70af-45f4-9d62-9713ea36f448)

🤔

**Comment by nesrak1** [comment]
> But if you just horizontally flip the texture before unswizzling (and horizontally flip again after), then you get this:

Yes, all unity images are always flipped vertically (all unity tools flip it again so it's right side up), so it makes sense you would need to do this. Is this C# that can be integrated into UABEA or just some external third party tool?

**Comment by fdwr** [comment]
> Is this C# that can be integrated into UABEA or just some external third party tool?

It's just some quick&dirty C++ code I wrote to play around with different [Morton patterns](https://fgiesen.wordpress.com/2011/01/17/texture-tiling-and-swizzling/). So it's not directly integratable, but here is the relevant nested tiling (which produces the element offsets in the image above):

```c++
    constexpr static CustomSwizzleItem items[] =
    {
        {MortonX, 2}, // 2x1
        {MortonY, 2}, // 2x2
        {MortonX, 2}, // 4x2
        {MortonY, 2}, // 4x4
        {MortonX, 2}, // 8x4
        {MortonY, 2}, // 8x8
    };
    uint32_t sourceOffset = CustomSwizzle(width, height, x, y, items, countof(items));
```

If you happen to already have some unswizzling code in there for other Morton patterns, you can fill in the tilings above.

For a faster approach (on systems where BMI2 is available with the pdep instruction, [Bmi2.ParallelBitDeposit](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.intrinsics.x86.bmi2.x64.parallelbitdeposit?view=net-9.0)), forming the offset within each 8x8 tile should just be 2 calls to `ParallelBitDeposit`. e.g. (note, I'm typing this out of my head directly on GitHub, not from a C# compiler, and so bugs may be present 😅):

```c#
uint widthInTiles = (width + 7) / 8;
const uint elementsPer8x8Tile = 8 * 8;

// ... for each y and x ...
uint linearByteAddress = (y * width + x) * bytesPerElement;
uint localOffsetInTile = ParallelBitDeposit(x & 7, 0b01010101) | ParallelBitDeposit(y & 7, 0b10101010);
uint tileOffset = ((y / 8) * widthInTiles + x / 8) * elementsPer8x8Tile;
uint swizzledByteAddress = (tileOffset + localOffsetInTile) * bytesPerElement;
// ... memcpy bytesPerElement from swizzledByteAddress to linearByteAddress ...
```

**Comment by nesrak1** [comment]
I'll look into it. It doesn't seem as different to some other swizzling code I made for other consoles before.

@eikun79 , if you're still around could you send the assets/bundle file+Texture2D+texture data? You previously sent an image, but from what I've seen in the past swizzling usually happens on the block level. In order to make sure I'm doing everything right, I want all of the extra metadata that comes with the texture. Here's how you can do it:

1. Open the file with the texture, press ctrl+a to select all assets, ctrl+click the texture you want to send over so it deselects it and click remove to delete everything else.
2. Save the modified bundle here.
3. Click view data on the remaining texture2d. If you expand the `m_StreamData` field and there is a file path there to a resS file, you will need to find where that resS file is. Otherwise, don't do anything else.

If the resS is small enough, you can probably just send the whole resS here. Otherwise, you'll need to extract just the bytes for this texture. For example, if it says `offset = 64` and `size = 4000`, in HxD you can press ctrl+e to select a region and put 64 and 4000 for offset and length.
![Image](https://github.com/user-attachments/assets/f7a97279-61a4-4448-8da1-2cfa5aed068e)

Then you could copy and paste the bytes into a new file to send along with the bundle.

**Comment by Der-Nogard** [comment]
Try this file [92c5723a4949e68430681a99c1a73d13.zip](https://github.com/user-attachments/files/19683845/92c5723a4949e68430681a99c1a73d13.zip)

**Comment by ToyBoyExtra** [comment]
> Recently, PS4 is also seeing an increasing number of games with Swizzle image files, such as Nintendo Switch Can you help me convert PS4 Swizzle files to Unsweezle as below?
> 
> [NotoSansJP-Regular SDF Atlas-sharedassets0.assets-590.zip](https://github.com/nesrak1/UABEA/files/13667489/NotoSansJP-Regular.SDF.Atlas-sharedassets0.assets-590.zip)
> 
> Example) [#185](https://github.com/nesrak1/UABEA/issues/185)

Hi friend. I have the same problem with PS4 textures as you. Did you solve it somehow? Did you manage to get a working tool to convert these textures? Thanks for the answer
